application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,165 @@
+public class WarnCommand extends SlashCommandAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(WarnCommand.class);
+    private static final String USER_OPTION = "user";
+    private static final String REASON_OPTION = "reason";
+    private static final String ACTION_VERB = "warn";
+    private final Database database;
+    private final Predicate<String> hasRequiredRole;
+
+    /**
+     * Creates a new Instance.
+     */
+    public WarnCommand(@NotNull Database database) {
+        super("warn", "warns the user", SlashCommandVisibility.GUILD);
+        this.database = database;
+
+        getData().addOption(OptionType.USER, USER_OPTION, "The user to warn", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "The reason for the warning", true);
+
+        hasRequiredRole = Pattern.compile(Config.getInstance().getSoftModerationRolePattern())
+            .asMatchPredicate();
+    }
+
+    private static @NotNull RestAction<Boolean> dmUser(long userId, @NotNull String reason,
+            @NotNull Guild guild, @NotNull SlashCommandEvent event) {
+        return event.getJDA()
+            .openPrivateChannelById(userId)
+            .flatMap(privateChannel -> event.replyEmbeds(new EmbedBuilder().setTitle("Warned")
+                .setDescription("You have been warned in " + guild.getName() + " for " + reason)
+                .setColor(Color.decode("#895FE8"))
+                .build()))
+            .mapToResult()
+            .map(Result::isSuccess);
+    }
+
+    private static @NotNull MessageEmbed sendFeedback(boolean hasSentDm, @NotNull User target,
+            @NotNull Member author, @NotNull String reason) {
+        String dmNoticeText = "";
+        if (!hasSentDm) {
+            dmNoticeText = "(Unable to send them a DM.)";
+        }
+        return ModerationUtils.createActionResponse(author.getUser(), ModerationUtils.Action.BAN,
+                target, dmNoticeText, reason);
+    }
+
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @Nullable Member target, @NotNull CharSequence reason, @NotNull Guild guild,
+            @NotNull Interaction event) {
+
+        // Member doesn't exist if attempting to ban a user who is not part of the guild.
+        if (target != null && !ModerationUtils.handleCanInteractWithTarget(ACTION_VERB, bot, author,
+                target, event)) {
+            return false;
+        }
+
+        if (!ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasBotPermissions(ACTION_VERB, Permission.KICK_MEMBERS, bot,
+                guild, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasAuthorPermissions(ACTION_VERB, Permission.KICK_MEMBERS,
+                author, guild, event)) {
+            return false;
+        }
+        return ModerationUtils.handleReason(reason, event);
```

Comment:
You just copied the logic from "ban" but warn is different. You are supposed to also be able to warn users who are not part of the guild anymore. For example after they received a temp ban.

Reply:
I am confused. Are you not also supposed to be able to ban users who are not part of the guild?